worker/src/main/java/gov/cms/ab2d/worker/adapter/bluebutton/ContractAdapterImpl.java

```diff
@@ -0,0 +1,186 @@
+                    }
+
+                    patientDTOs.add(patientDTO);
+                }
+            }
+        }
+
+        return GetPatientsByContractResponse.builder()
+                .contractNumber(contractNumber)
+                .patients(patientDTOs)
+                .build();
+    }
+
+
+    /**
+     * Given a contractNumber and month,
+     * calls BFD api to get a bundle of resources from which the patientIDs are extracted.
+     * If there are multiple pages available, fetches all pages.
+     *
+     * @param contractNumber
+     * @param month
+     * @return a list of PatientIds
+     */
+    private List<String> getPatientIdsForMonth(String contractNumber, Integer month) {
+        Bundle bundle = getBundle(contractNumber, month);
+        final List<String> patientIDs = extractPatientIDs(bundle);
+
+        while (bundle.getLink(Bundle.LINK_NEXT) != null) {
+            bundle = bfdClient.requestNextBundleFromServer(bundle);
+            patientIDs.addAll(extractPatientIDs(bundle));
+        }
+
+        return patientIDs;
+    }
+
+
+    /**
+     * given a contractNumber & a month, calls BFD API to find all patients active in the contract during the month
+     * @param contractNumber
+     * @param month
+     * @return a FHIR bundle of resources containing active patients
+     */
+    private Bundle getBundle(String contractNumber, int month) {
+        try {
+            return bfdClient.requestPartDEnrolleesFromServer(contractNumber, month);
+        } catch (Exception e) {
+            final Throwable rootCause = ExceptionUtils.getRootCause(e);
+            log.error("Error while calling for Contract-2-Bene API : {}", e.getMessage(), rootCause);
+            throw new RuntimeException(rootCause);
+        }
+    }
+
+    /**
+     * Given a Bundle, filters resources of type Patient and returns a list of patientIds
+     * @param bundle
+     * @return a list of patientIds
+     */
+    private List<String> extractPatientIDs(Bundle bundle) {
+        return bundle.getEntry().stream()
+                .map(Bundle.BundleEntryComponent::getResource)
+                .filter(resource -> resource.getResourceType() == ResourceType.Patient)
+                .map(resource -> (Patient) resource)
+                .map(patient -> extractPatientId(patient))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Given a patient, extract the patientId
+     * @param patient
+     * @return patientId if present, null otherwise
+     */
+    private String extractPatientId(Patient patient) {
+        return patient.getIdentifier().stream()
+                .filter(identifier -> isBeneficiaryId(identifier))
+                .map(Identifier::getValue)
+                .findFirst().orElse(null);
+    }
+
+    private boolean isBeneficiaryId(Identifier identifier) {
+        return identifier.getSystem().toLowerCase().endsWith("bene_id");
```

Comment:
It might be best to do a full comparison against the identifier's System value.

Reply:
Do you know what that value would be? 
I do not what that complete value for System would be.

I used the logic from here: 
https://github.com/CMSgov/ab2d/blob/master/optout/src/main/java/gov/cms/ab2d/optout/OptOutConverterServiceImpl.java#L124

Reply:
@spathiyil `https://bluebutton.cms.gov/resources/variables/bene_id`

Reply:
oh, I see. I should've used `equals("BENE_ID")` instead of`.endsWith("bene_id")`?

Reply:
`getSystem().equals("https://bluebutton.cms.gov/resources/variables/bene_id")`

Reply:
got it. thanks

Reply:
done